src/database.rs

```diff
@@ -84,59 +110,40 @@ pub trait Database: Send + AsDynDatabase + Any + ZalsaDatabase {
     #[cold]
     #[inline(never)]
     #[doc(hidden)]
-    fn zalsa_register_downcaster(&self) -> DatabaseDownCaster<dyn Database> {
-        self.zalsa().views().downcaster_for::<dyn Database>()
+    fn zalsa_register_upcaster(&self) -> &DatabaseUpCaster<dyn Database> {
+        self.zalsa().views().upcaster_for::<dyn Database>()
         // The no-op downcaster is special cased in view caster construction.
     }
 
     #[doc(hidden)]
     #[inline(always)]
-    unsafe fn downcast(db: &dyn Database) -> &dyn Database
+    fn upcast(&self) -> &dyn Database
     where
         Self: Sized,
     {
         // No-op
-        db
+        self
     }
 }
 
 /// Upcast to a `dyn Database`.
 ///
-/// Only required because upcasts not yet stabilized (*grr*).
+/// Only required because upcasting does not work for unsized generic parameters.
```

Comment:
I am quite annoyed by this but that's a limitation of Rust, a `&T` of `T: ?Sized + SomeTrait` cannot be coerced to a `&dyn SomeTrait` (or a supertrait) due to `T` possibly being an unrelated unsized type like `str`

Reply:
No, specifically this is used solely for `attach` which needs to take a dyn database of some form.